src/stream/src/executor/barrier_align.rs

```diff
@@ -60,8 +88,14 @@ pub async fn barrier_align(
                 // left stream end, passthrough right chunks
                 while let Some(msg) = right.next().await {
                     match msg? {
-                        Message::Watermark(_) => {
-                            todo!("https://github.com/risingwavelabs/risingwave/issues/6042")
+                        Message::Watermark(just_received_watermark) => {
+                            right_cached_watermarks.push_back(just_received_watermark);
```

Comment:
We should group `Watermark`s by the column index. They should be handled independently.

Reply:
> We should group `Watermark`s by the column index. They should be handled independently.

fixed